tests/libcxx/host/host.cpp

```diff
@@ -12,202 +12,200 @@
 #include <cstring>
 #include <map>
 #include <thread>
-#include "args.h"
-#include "ocalls.h"
+#include "libcxx_u.h"
 #include "threadArgs.h"
 
 // Host maintains a map of enclave key to host thread ID
-static std::map<uint64_t, pthread_t> enclave_host_id_map;
-static std::atomic_flag _host_lock = ATOMIC_FLAG_INIT;
+static std::map<uint64_t, pthread_t> _enclave_host_id_map;
+static atomic_flag_lock _host_lock;
 
-void Test(oe_enclave_t* enclave)
+void test(oe_enclave_t* enclave)
 {
-    Args args;
-    args.ret = 1;
-    args.test = NULL;
-    oe_result_t result = oe_call_enclave(enclave, "Test", &args);
+    int ret = 1;
+    char test_name[STRLEN];
+    oe_result_t result = enc_test(enclave, &ret, test_name);
     OE_TEST(result == OE_OK);
 
-    if (args.ret == 0)
+    if (ret == 0)
     {
-        printf("PASSED: %s\n", args.test);
+        printf("PASSED: %s\n", test_name);
     }
     else
     {
-        printf("FAILED: %s (ret=%d)\n", args.test, args.ret);
+        printf("FAILED: %s (ret=%d)\n", test_name, ret);
         abort();
     }
 }
 
-OE_OCALL void ocall_exit(uint64_t arg)
+void host_exit(int arg)
 {
-    exit(static_cast<int>(arg));
+    exit(arg);
 }
 
-void* EnclaveThread(void* args)
+struct thread_args_t
 {
-    ThreadArgs* thread_args = (ThreadArgs*)args;
-    pthread_t host_thread_id = pthread_self();
+    oe_enclave_t* enclave;
+    uint64_t enc_key;
+};
+
+void* host_enclave_thread(void* args)
+{
+    thread_args_t* thread_args = reinterpret_cast<thread_args_t*>(args);
+    // need to cache the values for enc_key and enclave now before _host_lock
+    // is unlocked after assigning the thread_id to the _enclave_host_id_map
+    // because args is a local variable in the calling method which may exit at
+    // any time after _host_lock is unlocked which may cause a segfault
+    uint64_t enc_key = thread_args->enc_key;
```

Comment:
Unless it is declared volatile, the compiler is free to optimize away these local variables and assignments.
In this case, the compiler will emit the loads here as expected due to the atomic_lock below; however that is a bit subtle to infer.

Reply:
In this case, thread_args is a pointer to a local stack variable from host_create_thread (line 82).  host_create_thread waits in a spin lock (line 103) waiting for the _enclave_host_id_map[enc_key] to be set on line  of this method.  After line 65, it is no longer safe to reference thread_args within this method because host_create_thread may exit at any time after that point.  This method creates local values for enc_key, and enclave because the memory must be assumed to be invalid after line 65.  In this case, the values are not volatile and do not change, but the memory location is temporary.
The prior version of this code allocated the memory on the heap which avoided this issue but never de-allocated the memory.